modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/DistributionZoneRebalanceEngineV2.java

```diff
@@ -0,0 +1,287 @@
+     * @param catalogVersion Catalog version.
+     * @return The future, which completes when the all metastore updates done.
+     */
+    private CompletableFuture<Void> recalculateAssignmentsAndScheduleRebalance(
+            CatalogZoneDescriptor zoneDescriptor,
+            long causalityToken,
+            int catalogVersion
+    ) {
+        return distributionZoneManager.dataNodes(causalityToken, catalogVersion, zoneDescriptor.id())
+                .thenCompose(dataNodes -> {
+                    if (dataNodes.isEmpty()) {
+                        return nullCompletedFuture();
+                    }
+
+                    return triggerPartitionsRebalanceForZone(
+                            causalityToken,
+                            zoneDescriptor,
+                            dataNodes
+                    );
+                });
+    }
+
+    private CompletableFuture<Void> triggerPartitionsRebalanceForZone(
+            long revision,
+            CatalogZoneDescriptor zoneDescriptor,
+            Set<String> dataNodes
+    ) {
+        CompletableFuture<?>[] partitionFutures = ZoneRebalanceUtil.triggerZonePartitionsRebalance(
+                zoneDescriptor,
+                dataNodes,
+                revision,
+                metaStorageManager
+        );
+
+        // This set is used to deduplicate exceptions (if there is an exception from upstream, for instance,
+        // when reading from MetaStorage, it will be encountered by every partition future) to avoid noise
+        // in the logs.
+        Set<Throwable> unwrappedCauses = ConcurrentHashMap.newKeySet();
+
+        for (int partId = 0; partId < partitionFutures.length; partId++) {
+            int finalPartId = partId;
+
+            partitionFutures[partId].exceptionally(e -> {
+                Throwable cause = ExceptionUtils.unwrapCause(e);
+
+                if (unwrappedCauses.add(cause)) {
+                    // The exception is specific to this partition.
+                    LOG.error(
+                            "Exception on updating assignments for [zone={}, partition={}]",
+                            e,
+                            zoneInfo(zoneDescriptor), finalPartId
+                    );
+                } else {
+                    // The exception is from upstream and not specific for this partition, so don't log the partition index.
+                    LOG.error(
+                            "Exception on updating assignments for [zone={}]",
+                            e,
+                            zoneInfo(zoneDescriptor)
+                    );
+                }
+
+                return null;
+            });
+        }
+
+        return allOf(partitionFutures);
+    }
+
+    private static String zoneInfo(CatalogZoneDescriptor zoneDescriptor) {
+        return zoneDescriptor.id() + "/" + zoneDescriptor.name();
+    }
+
+    static CompletableFuture<Set<Assignment>> calculateZoneAssignments(
```

Comment:
- triggerPartitionsRebalanceForZone is a part of DistributionZoneRebalanceEngineV2 while triggerZonePartitionsRebalance is a part of ZoneRebalanceUtil. Why?
- The fact that we reorder the words in the method names is confusing. I mean the aforementioned methods.
- calculateZoneAssignments is a part of DistributionZoneRebalanceEngineV2 while it's never used from within the class.

Reply:
refactored